### PR TITLE
Fluorescence CLI

### PR DESCRIPTION
**Pull request by talonchandler**
@ziw-liu I'm requesting your review of this fairly large PR. I have tested this PR on hummingbird, and I'd like to prioritize getting this merged into `0.4.0dev` then `main` so that we can start integration testing for the `0.4.0` release. 

I'm particularly interested in your suggestions related to (1) the new `/examples/` folder for the new config file formats, (2) the new CLI names and simplified `recOrder reconstruct` CLI, and (3) the new `cli.settings.py` file. 

**Summary of major changes**:
- new fluorescence CLI functionality
- a single configuration file design instead of two configuration files (one for `compute-transfer-function` and one for `apply-inverse-transfer-function`
- reworked pydantic models for configuration files (now supported birefringence, phase, birefringence + phase, and fluorescence)
- reworked `examples/` folder demonstrating new CLI

Minor changes:
- high-level tests of CLI
- new CLI names: `compute-tf` and `apply-inv-tf`. These shortened names are for the CLI only. 
- basic validation of CLI inputs
- automatically generated example configuration files

Earlier list of TODOs:
- [x] finish high-level `recorder reconstruct` CLI and test
- [x] integration testing at the microscope
- [x] removal of example scripts with documentation (now replaced by `waveorder` scripts and CLI + example configurations)
- [x] documentation of configuration file generation

@edyoshikun I'm tagging you as a reviewer, but we'll likely merge before you're back. We'll make sure this is tested against your upcoming use cases before the 0.4.0 release.

### CODE REVIEW

**Comment by edyoshikun** [comment]
I know this has been a lot of work but it's great to see it come together! I am wondering if it makes sense to have the `apply_inverse_transfer_functions` for all recons to be separate functions.
We can do it for another PR post merging `0.4.0dev`. I attempted something in #380 at least for the birefringence reconstruction, which I pushed for you to have the code just in case while I was gone.

File: recOrder/cli/main.py [hunk 1]
```diff
@@ -14,5 +12,5 @@ def cli():
 
 cli.add_command(view)
 cli.add_command(reconstruct)
-cli.add_command(compute_transfer_function)
-cli.add_command(apply_inverse_transfer_function)
+cli.add_command(compute_tf)
```
Comment (edyoshikun) [inline recOrder/cli/main.py:15]: thanks for keeping this!

File: recOrder/cli/apply_inverse_transfer_function.py [hunk 2]
```diff
@@ -135,10 +97,8 @@ def apply_inverse_transfer_function_cli(
     tczyx_data = torch.tensor(input_dataset.data, dtype=torch.float32)
```
Comment (edyoshikun) [inline recOrder/cli/apply_inverse_transfer_function.py:97]: when I ran this with the `uint16` datasets, I was getting an error that torch did not support that format and had to open `input_dataset.data.astype(np.float32)` instead.

Comment (edyoshikun) [reply recOrder/cli/apply_inverse_transfer_function.py:97]: I noticed this when the datasets datatype when using `iohub info -v ./dataset.zarr`

File: recOrder/cli/apply_inverse_transfer_function.py [hunk 3]
```diff
@@ -280,17 +233,15 @@ def apply_inverse_transfer_function_cli(
...
                     **biref_inverse_dict,
                 )
 
                 reconstructed_parameters_3d = inplane_oriented_thick_pol3d.apply_inverse_transfer_function(
                     tczyx_data[time_index],
                     intensity_to_stokes_matrix,
-                    wavelength,
-                    cyx_no_sample_data,
+                    cyx_no_sample_data=cyx_no_sample_data,
                     project_stokes_to_2d=False,
```
Comment (edyoshikun) [inline recOrder/cli/apply_inverse_transfer_function.py:245]: should this be consistent to previous?
(ndim==2)

File: recOrder/cli/apply_inverse_transfer_function.py [hunk 4]
```diff
@@ -342,8 +290,7 @@ def apply_inverse_transfer_function_cli(
...
```
Comment (edyoshikun) [inline recOrder/cli/apply_inverse_transfer_function.py:294]: (ndim==2)

File: recOrder/cli/apply_inverse_transfer_function.py (hunk 3 above)
Comment (talonchandler) [reply recOrder/cli/apply_inverse_transfer_function.py:245]: Sorry, I'm having trouble figuring out what you mean here. Can you expand on this comment and the next?

File: recOrder/cli/apply_inverse_transfer_function.py (hunk 2 above)
Comment (talonchandler) [reply recOrder/cli/apply_inverse_transfer_function.py:97]: Did you run into this problem on this branch as well? I'm testing with `np.uint16` here, but you may have seen a different issue. If this is still an issue, can you point to an example dataset?

https://github.com/mehta-lab/recOrder/blob/ea9d4c976bdb2b5c019aec5741563ca14d1259e8/recOrder/tests/cli_tests/test_apply_inv_tf.py#L20

**Review by edyoshikun — COMMENTED** [review]
Other than the comments, which are minor fixes, this LGTM! I tested it with the update on `alg-dev` [waveorder #131](https://github.com/mehta-lab/waveorder/pull/131) and the dataset from zenodo.

File: examples/birefringence-and-phase.yml [hunk 5]
```diff
@@ -0,0 +1,30 @@
...
+- State3
+reconstruction_dimension: 3
+birefringence:
+  transfer_function:
+    swing: 0.1
+  apply_inverse:
+    wavelength_illumination: 0.532
+    background_path: ''
+    remove_estimated_background: false
+    orientation_flip: false
```
Comment (edyoshikun) [inline examples/birefringence-and-phase.yml:14]: not sure if I understand this parameter. Is the orientation flip doing a mirror or a 180 deg rotation?

Comment (edyoshikun) [reply examples/birefringence-and-phase.yml:14]: I realized these are not implemented yet so we can change the naming afterwards

File: recOrder/cli/apply_inverse_transfer_function.py (hunk 3 above)
Comment (talonchandler) [reply recOrder/cli/apply_inverse_transfer_function.py:262]: This block can only be reached `recon_dim == 3` (similar for `recon_dim == 2`), so I think I'm already using `recon_dim` in the way you're suggesting. 

The birefringence + phase reconstructions in 2D and 3D have relatively intricate logic wrt the waveorder calls, so I don't think there's a dramatic simplification here (I'm open to being wrong though!).

File: recOrder/cli/apply_inverse_transfer_function.py (hunk 4 above)
Comment (talonchandler) [reply recOrder/cli/apply_inverse_transfer_function.py:311]: See comment above...feel free to reopen.

File: recOrder/cli/reconstruct.py [hunk 6]
```diff
@@ -1,7 +1,44 @@
...
+@output_dataset_option(default="./reconstruction.zarr")
+def reconstruct(input_data_path, config_path, output_path):
+    """Reconstruct a dataset using a configuration file. This is a
+    convenience function for a `compute-tf` call followed by a `apply-inv-tf`
+    call.
+
+    See /examples/settings/ for example configuration files.
+
+    Example usage:\n
+    $ recorder reconstruct input.zarr -c /examples/settings/birefringence.yml -o output.zarr
```
[the remark this answers is not in the document]
Comment (talonchandler) [reply recOrder/cli/reconstruct.py:29]: Great catch!

**Comment by talonchandler** [comment]
Thanks @edyoshikun. Merging into `0.4.0dev`

**Pull request merged**